core/src/global.rs

```diff
@@ -224,35 +222,22 @@ pub fn create_pow_context<T>(
 	max_sols: u32,
 ) -> Result<Box<dyn PoWContext>, pow::Error> {
 	let chain_type = get_chain_type();
-	match chain_type {
-		// Mainnet has Cuckaroo{,d,m,z}29 for AR and Cuckatoo31+ for AF
-		ChainTypes::Mainnet if edge_bits > 29 => new_cuckatoo_ctx(edge_bits, proof_size, max_sols),
-		ChainTypes::Mainnet if valid_header_version(height, HeaderVersion(4)) => {
-			new_cuckarooz_ctx(edge_bits, proof_size)
-		}
-		ChainTypes::Mainnet if valid_header_version(height, HeaderVersion(3)) => {
-			new_cuckaroom_ctx(edge_bits, proof_size)
-		}
-		ChainTypes::Mainnet if valid_header_version(height, HeaderVersion(2)) => {
-			new_cuckarood_ctx(edge_bits, proof_size)
-		}
-		ChainTypes::Mainnet => new_cuckaroo_ctx(edge_bits, proof_size),
-
-		// Same for Testnet
-		ChainTypes::Testnet if edge_bits > 29 => new_cuckatoo_ctx(edge_bits, proof_size, max_sols),
-		ChainTypes::Testnet if valid_header_version(height, HeaderVersion(4)) => {
-			new_cuckarooz_ctx(edge_bits, proof_size)
-		}
-		ChainTypes::Testnet if valid_header_version(height, HeaderVersion(3)) => {
-			new_cuckaroom_ctx(edge_bits, proof_size)
-		}
-		ChainTypes::Testnet if valid_header_version(height, HeaderVersion(2)) => {
-			new_cuckarood_ctx(edge_bits, proof_size)
+	if chain_type == ChainTypes::Mainnet || chain_type == ChainTypes::Testnet {
+		// Mainnet and Testnet have Cuckatoo31+ for AF and Cuckaroo{,d,m,z}29 for AR
+		if edge_bits > 29 {
```

Comment:
Consider making this explicit on `header_version()` also?

```
if header_version(height) >= 5 || edge_bits > 29 {
   ...
}
```

Would we then want to explicitly fail if we actually had `edge_bits <= 29` after HF4?

Reply:
Asking for a 29 context past HF4 should result in a clear error such as no_cuckaroo_ctx() rather than a dubious CuckAToo29 context, which is not an outright error, but otherwise quite invalid.

Reply:
Yes absolutely - just wondering if there is a more explicit way of implementing this, failing earlier in the code with a `no_cuckatoo_ctx()`.